zmkfirmware/zmk
# feat(split): sync central & peripherals last activity timing

Sync central last activity timings to all devices, by having the central emit how long it's inactive, and the peripheral(s) using it to adjust the local last activity time.

Before this commit, key presses on a peripheral keep the central awake, but not vice versa. With this commit, key presses on the central (or hypothetical peripheral B) can keep peripheral A awake.

This is done by:
1. Adding a new `SYNC_ACTIVITY` GATT characteristic for central to sync its inactive time to the other peripheral(s).
2. Central's `activity.c` broadcasting the inactive time at a regular `ZMK_SPLIT_SYNC_LAST_ACTIVITY_INTERVAL_MS` interval
3. Peripheral's `service.c` receiving the data and sending it to its `activity.c`
4. Peripheral's `activity.c` determining the new `activity_last_uptime` value using the inactive duration as a relative difference.

Additionally:
- `central.c` is updated to sync the timers upon a new BLE connection, so new devices don't need to wait for the full interval to get synced.
- If a peripheral is in IDLE mode when the central comes online, it'll get synced to ACTIVE as well.

Cons:
- This solution doesn't handle cases where the peripheral(s) is already in deep sleep, since BLE is turned off. However, the sync-ing can prevent peripheral(s) from going into deep sleep if the central is used.
- If `ZMK_SPLIT_SYNC_SLEEP_TIMERS_INTERVAL_MS` > `ZMK_IDLE_TIMEOUT`, there can be situations where the peripheral(s) goes into IDLE before receiving the next sync from central. The original default interval was 5 minutes, primarily to prevent unnecessary deep sleeps, but it would run into the above case.
- This is a time-based periodic sync, so it won't be as accurate as an event-based sync, where each key press is sync-ed to the peripheral(s). The latter would potentially take up more of the BLE channel bandwidth. Maybe this accurate mode can be toggled as an option?

Test steps
---------
1. Build and flash both sides to the target keyboard, ideally with indicators such as backlight, battery, RGB underglow, displays, to detect when one side has gone to idle/active.
2. Allow the peripheral to go to idle state (default timeout: 30s)
3. Press any key on the central that causes it to switch to active. The peripheral should go to active state, **not immediately**, but within 30s.

Local test
------------
I added log lines to check the output, that the values are sent and received correctly on both ends. However, if the sync interval is set to a low value (30 seconds), the log lines would end up spamming the console, so I opt'ed to remove them. Let me know if logs can be introduced more nicely, or if there's a better way to check that the code works for the context of reviewing this pull request!


Raised as an issue in https://github.com/zmkfirmware/zmk/issues/2408
